### PR TITLE
fix: allow replacement tenure-start block with pre-commit

### PR DESCRIPTION
**Pull request by brice-stacks**
On mainnet, we have had stalls due to the fact that a signer would not accept a replacement tenure start block if it had already pre-committed (not signed!) to a different tenure start block. This scenario would not resolve until the next burn block.

Fixes: stx-labs/core-epics#336

### Checklist

- [x] Test coverage for new or modified code paths
- [ ] For new Clarity features or consensus changes, add property tests (see [`docs/property-testing.md`](/docs/property-testing.md))
- [x] Changelog fragment(s) or "no changelog" label added (see [`changelog.d/README.md`](changelog.d/README.md))
- [ ] Required documentation changes (e.g., [`rpc/openapi.yaml`](/docs/rpc/openapi.yaml) for RPC endpoints, [`event-dispatcher.md`](/docs/event-dispatcher.md) for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo

### CODE REVIEW

**Comment by brice-stacks** [comment]
I think my changes here did actually cause that new flakiness (it's not just already flaky). I'm taking a deeper look.

File: stacks-signer/src/chainstate/tests/v2.rs [hunk 1]
```diff
@@ -745,3 +745,105 @@ fn check_tenure_change_rejects_when_locally_accepted_block_exists() {
...
+
+/// Test that a tenure change proposal is accepted when the only block in the
+/// same tenure is one the signer has merely pre-committed to (but not signed).
+///
+/// This is a regression test for a stall seen on mainnet: a signer
+/// pre-committed to a tenure-start block that never reached consensus, and
+/// then rejected the miner's replacement tenure-start block with
+/// `DuplicateBlockFound` because the pre-committed block was treated as
+/// accepted. A pre-commit carries no signature, so it is safe to accept a
+/// competing proposal at the same height.
```
Comment (aaronb-stacks) [inline stacks-signer/src/chainstate/tests/v2.rs:757]: Can you add a test that covers the negative case: test that a blockfound tenure change proposal is rejected when the only block in the same tenure is one that the signer has locally signed?

**Review by aaronb-stacks — COMMENTED** [review]
This looks great to me, just had a comment about an additional test.

File: stacks-signer/src/v0/signer.rs [hunk 2]
```diff
@@ -1156,6 +1156,55 @@ impl Signer {
...
+                // consensus, proposal evaluation already allows blocks that do not confirm it,
+                // so allow signing the replacement too -- otherwise the tenure stalls until the
+                // next sortition.
+                let signature_is_fresh = last_signed.signed_self.is_some_and(|signed_time| {
+                    signed_time.saturating_add(
+                        self.proposal_config
+                            .tenure_last_block_proposal_timeout
+                            .as_secs(),
+                    ) > get_epoch_time_secs()
+                });
```
[the remark this answers is not in the document]
Comment (brice-stacks) [reply stacks-signer/src/v0/signer.rs]: I've patched this. @aaronb-stacks please take a close look at stacks-signer/src/v0/signer.rs:1159-1234 to both verify the policy I am trying to implement, and the logic I used to do so. If you have any ideas for how to do it cleaner, I am open to suggestions, but this is the best I've come up with.

File: stacks-signer/src/chainstate/tests/v2.rs (hunk 1 above)
Comment (brice-stacks) [reply stacks-signer/src/chainstate/tests/v2.rs:757]: Thanks for the patch. I built 2 more tests off of that framework.

File: stacks-signer/src/v0/signer.rs [hunk 3]
```diff
@@ -1156,6 +1156,82 @@ impl Signer {
...
+        ) {
+            Ok(signed_blocks) => signed_blocks,
+            Err(e) => {
+                warn!("{self}: Failed to query the signed blocks in the tenure. Refusing to sign block {block_hash}: {e:?}");
+                return;
+            }
+        };
+        // Find the freshest conflicting block (i.e. the one with the most recent endorsement) and
+        // check if it is still fresh. This list could contain the block we are about to sign, so
+        // we filter that out (it can't conflict with itself).
```
Comment (aaronb-stacks) [inline stacks-signer/src/v0/signer.rs]: Hmm -- I don't really like this, for a couple reasons, but mostly because it involves an unbounded SQL SELECT.

Would something like the following make sense?

```rust
let mut highest_maybe_conflict = signer_db.get_last_signed_block();
let has_conflicting_ancestor = loop {
  if highest_maybe_conflict.block_height < info.block.block_height {
     break false;
  }
  if highest_maybe_conflict.block_height == info.block.block_height &&
     highest_maybe_conflict.signer_sig_hash == info.signer_sig_hash {
     break false;
  }
  if highest_maybe_conflict.is_globally_accepted {
     break true;
  }
  if highest_maybe_conflict.is_locally_accepted && not_timed_out {
     break true;
  }
  highest_maybe_conflict = signer_db.get_highest_signed_block_less_than(
     highest_maybe_conflict.block_height
  );
}

if has_conflicting_ancestor {
    return block reject;
}
```

Basically, we search backwards from the highest block.

File: stacks-signer/src/v0/signer.rs [hunk 4]
```diff
@@ -1156,6 +1156,82 @@ impl Signer {
...
+                    "conflicting_block_height" => conflict.block.header.chain_length,
+                );
+                return;
+            }
+
+            // The freshest conflict has timed out, but if it was globally accepted, we still
+            // cannot sign the replacement unless it is no longer considered canonical by the node
+            // (a Bitcoin reorg may have orphaned it). If the node cannot be reached, assume the
+            // proposed block is higher, as proposal evaluation does.
+            match stacks_client.get_tenure_tip(&block_info.block.header.consensus_hash) {
```
Comment (aaronb-stacks) [inline stacks-signer/src/v0/signer.rs]: Is this `get_tenure_tip` check necessary or is it just a defense-in-depth? I think the block proposal validation endpoint already performs this check -- and invoking the RPC endpoint again will add some overhead that I'd be a little worried about.

Comment (aaronb-stacks) [reply stacks-signer/src/v0/signer.rs]: Nevermind -- I see the scenario where this would be necessary.

File: stacks-signer/src/v0/signer.rs (hunk 3 above)
Comment (brice-stacks) [reply stacks-signer/src/v0/signer.rs]: I switched to two queries instead: (1) retrieves a fresh conflicting block, and (2) — only if that's empty — retrieves any conflicting block. Both are LIMIT 1 with no ORDER BY, so they stop at the first match, and the query plan shows them fully served by the blocks_consensus_hash_state_height index — no block bodies are ever loaded. I think that's cheaper than the iteration in Rust, which would still visit every stale block in the recovery case. 5abd8cb1c6

**Review by aaronb-stacks — DISMISSED** [review]

File: stacks-signer/src/v0/signer.rs [hunk 5]
```diff
@@ -1156,6 +1156,87 @@ impl Signer {
...
+        };
+        if let Some(conflict) = fresh_conflict {
+            warn!(
+                "{self}: Reached the pre-commit threshold for a block, but we have recently signed or accepted a different block at the same or higher height in this tenure. Refusing to sign.";
+                "signer_signature_hash" => %block_hash,
+                "block_height" => block_info.block.header.chain_length,
+                "conflicting_signer_signature_hash" => %conflict.signer_signature_hash,
+                "conflicting_block_height" => conflict.stacks_height,
+            );
+            return;
```
Comment (federico-stacks) [inline stacks-signer/src/v0/signer.rs:1192]: sanity-check: I’m wondering whether the following scenario is now possible:

1. The new guard returns early, leaving the block in the `PreCommitted` state with `valid = true`.
2. After a signature timeout, the miner resubmits the same proposal.
3. `should_reevaluate_block()` calls `determine_response()`, which treats `valid == true` as accepted and creates a fresh signature.

Should a resubmitted `PreCommitted` block re-enter `handle_block_pre_commit()` before it can be signed? That would also give the timeout/tenure-tip branch another opportunity to run after a fresh conflict initially caused an early return.

Comment (brice-stacks) [reply stacks-signer/src/v0/signer.rs:1192]: Good call.

Comment (brice-stacks) [reply stacks-signer/src/v0/signer.rs:1192]: Fixed in 678158707abd98beb6004e23fadd7898629b2d75

File: stacks-signer/src/v0/signer.rs [hunk 6]
```diff
@@ -1194,6 +1276,25 @@ impl Signer {
...
+                    "block_id" => %block_info.block.block_id(),
+                    "block_height" => block_info.block.header.chain_length,
+                    "burn_height" => block_proposal.burn_height,
+                    "consensus_hash" => %block_info.block.header.consensus_hash
+                );
+                self.send_block_pre_commit(signer_signature_hash.clone());
+                let address = self.stacks_address.clone();
+                self.handle_block_pre_commit(stacks_client, &address, &signer_signature_hash);
+                return false;
+            }
```
Comment (federico-stacks) [inline stacks-signer/src/v0/signer.rs:1297]: Just a comment on this change introduced here https://github.com/stacks-network/stacks-core/pull/7429#discussion_r3674540711: if we've already refused it and the conflict is still fresh, we'll end up re-announcing it via `send_block_pre_commit`, effectively sending a divergent signal to peers ("I will sign this"). That shouldn't be an issue, though, since signers that have already recorded our pre-commit will treat the duplicate as a no-op (maybe similar reasoning could apply to the other caller `handle_block_validate_ok` that uses the pair `send_block_pre_commit`/`handle_block_pre_commit`).

**Review by federico-stacks — APPROVED** [review]

**Review by aaronb-stacks — APPROVED** [review]
